File: 2017/js/day5.js
window.AdventOfCode.Day5 = function( input )
{
	let part2offsets = input.split( '\n' ).map( ( num ) => parseInt( num, 10 ) );
	
	let part1offsets = [...part2offsets];
	let offset = 0;
	let part1 = 0;
	let part2 = 0;
	
	do
	{
		part1++;
		
		const nextOffset = part1offsets[ offset ];
		
		part1offsets[ offset ]++;
		
		offset += nextOffset;
	}
	while( offset > -1 && offset < part1offsets.length );
	
	offset = 0;
	
	do
	{
		part2++;
		
		const nextOffset = part2offsets[ offset ];
		
		if( nextOffset >= 3 )
		{
			part2offsets[ offset ]--;
		}
		else
		{
			part2offsets[ offset ]++;
		}
		
		offset += nextOffset;
	}
	while( offset > -1 && offset < part2offsets.length );
	
	return [ part1, part2 ];
};

File: 2024/Answers/Solutions/Day5.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;

namespace AdventOfCode;

[Answer(5)]
public class Day5 : IAnswer
{
	public Solution Solve(string input)
	{
		var part1 = 0;
		var part2 = 0;
		var span = input.AsSpan();
		var split = span.IndexOf("\n\n");
		var dependents = new Dictionary<int, List<int>>();

		foreach (var line in span[..split].EnumerateLines())
		{
			var b = ParseTwoInt(line[0..]);
			var a = ParseTwoInt(line[3..]);

			if (!dependents.TryGetValue(b, out var list))
			{
				list = [];
				dependents[b] = list;
			}

			list.Add(a);
		}

		var pages = new List<int>(32);

		foreach (var line in span[(split + 2)..].EnumerateLines())
		{
			pages.Clear();

			foreach (var page in line.Split(','))
			{
				var number = ParseTwoInt(line[page].ToString());
				pages.Add(number);
			}

			var valid = true;

			for (var i = 0; i < pages.Count; i++)
			{
				if (!dependents.TryGetValue(pages[i], out var deps))
				{
					continue;
				}

				foreach (var dep in deps)
				{
					var pos = pages.IndexOf(dep, 0, i);

					if (pos > -1)
					{
						valid = false;
					}
				}
			}

			if (valid)
			{
				part1 += pages[pages.Count / 2];
				continue;
			}

			// part 2
			do
			{
				valid = true;

				for (var i = 0; i < pages.Count; i++)
				{
					if (!dependents.TryGetValue(pages[i], out var deps))
					{
						continue;
					}

					foreach (var dep in deps)
					{
						var pos = pages.IndexOf(dep, 0, i);

						if (pos > -1)
						{
							(pages[i], pages[pos]) = (pages[pos], pages[i]);
							valid = false;
						}
					}
				}
			}
			while (!valid);

			part2 += pages[pages.Count / 2];
		}

		return new(part1.ToString(), part2.ToString());
	}


	static int ParseTwoInt(ReadOnlySpan<char> line) => 10 * (line[0] - '0') + (line[1] - '0');
}
